src/io/iter_image_recordio_2.cc

```diff
@@ -0,0 +1,532 @@
+#include <dmlc/io.h>
+#include <dmlc/omp.h>
+#include <dmlc/common.h>
+#include <dmlc/timer.h>
+#include <type_traits>
+#include "./image_recordio.h"
+#include "./image_augmenter.h"
+#include "./image_iter_common.h"
+#include "./inst_vector.h"
+
+namespace mxnet {
+namespace io {
+// parser to parse image recordio
+template<typename DType>
+class ImageRecordIOParser2 {
+ public:
+  // initialize the parser
+  inline void Init(const std::vector<std::pair<std::string, std::string> >& kwargs);
+
+  // set record to the head
+  inline void BeforeFirst(void) {
+    if (batch_param_.round_batch == 0 || !overflow)
+      return source_->BeforeFirst();
+    else
+      overflow = false;
+  }
+  // parse next set of records, return an array of
+  // instance vector to the user
+  inline bool ParseNext(DataBatch *out);
+
+ private:
+  inline void ParseChunk(dmlc::InputSplit::Blob * chunk);
+  inline void CreateMeanImg(void);
+
+  // magic number to seed prng
+  static const int kRandMagic = 111;
+  static const int kRandMagicNormalize = 0;
+  /*! \brief parameters */
+  ImageRecParserParam param_;
+  ImageRecordParam record_param_;
+  BatchParam batch_param_;
+  ImageNormalizeParam normalize_param_;
+  PrefetcherParam prefetch_param_;
+  #if MXNET_USE_OPENCV
+  /*! \brief augmenters */
+  std::vector<std::vector<std::unique_ptr<ImageAugmenter> > > augmenters_;
+  #endif
+  /*! \brief random samplers */
+  std::vector<std::unique_ptr<common::RANDOM_ENGINE> > prnds_;
+  common::RANDOM_ENGINE rnd_;
+  /*! \brief data source */
+  std::unique_ptr<dmlc::InputSplit> source_;
+  /*! \brief label information, if any */
+  std::unique_ptr<ImageLabelMap> label_map_;
+  /*! \brief temporary results */
+  std::vector<InstVector<DType>> temp_;
+  /*! \brief temp space */
+  mshadow::TensorContainer<cpu, 3> img_;
+  /*! \brief internal instance order */
+  std::vector<std::pair<unsigned, unsigned> > inst_order_;
+  unsigned inst_index_;
+  /*! \brief internal counter tracking number of already parsed entries */
+  unsigned n_parsed_;
+  /*! \brief overflow marker */
+  bool overflow;
+  /*! \brief unit size */
+  std::vector<size_t> unit_size_;
+  /*! \brief mean image, if needed */
+  mshadow::TensorContainer<cpu, 3> meanimg_;
+  // whether mean image is ready.
+  bool meanfile_ready_;
+};
+
+template<typename DType>
+inline void ImageRecordIOParser2<DType>::Init(
+    const std::vector<std::pair<std::string, std::string> >& kwargs) {
+#if MXNET_USE_OPENCV
+  // initialize parameter
+  // init image rec param
+  param_.InitAllowUnknown(kwargs);
+  record_param_.InitAllowUnknown(kwargs);
+  batch_param_.InitAllowUnknown(kwargs);
+  normalize_param_.InitAllowUnknown(kwargs);
+  prefetch_param_.InitAllowUnknown(kwargs);
+  n_parsed_ = 0;
+  overflow = false;
+  rnd_.seed(kRandMagic + record_param_.seed);
+  int maxthread, threadget;
+  #pragma omp parallel
+  {
+    // be conservative, set number of real cores
+    maxthread = std::max(omp_get_num_procs() / 2 - 1, 1);
```

Comment:
Yes, and the number of threads is a min(this, parameter). Original iterator had the same logic.